concordium-std/src/impls.rs

```diff
@@ -251,46 +391,1129 @@ impl Write for ContractState {
+    }
+
+    /// Remove a key from the map, returning the value at the key if the key was
+    /// previously in the map.
+    ///
+    /// *Caution*: If `V` is a [StateBox], [StateMap], then it is
+    /// important to call [`Deletable::delete`] on the value returned when
+    /// you're finished with it. Otherwise, it will remain in the contract
+    /// state.
+    #[must_use]
+    pub fn remove_and_get(&mut self, key: &K) -> Option<V> {
+        let key_bytes = self.key_with_map_prefix(key);
+        // Unwrapping is safe because iter() holds a reference to the stateset.
+        let entry_raw = self.state_api.entry(key_bytes);
+        match entry_raw {
+            EntryRaw::Vacant(_) => None,
+            EntryRaw::Occupied(mut occ) => {
+                // Unwrapping safe in high-level API.
+                let old_value =
+                    V::deserial_with_state(&self.state_api, occ.get_mut()).unwrap_abort();
+                let _existed = self.state_api.delete_entry(occ.state_entry);
+                Some(old_value)
+            }
+        }
+    }
+
+    /// Remove a key from the map.
+    /// This also deletes the value in the state.
+    pub fn remove(&mut self, key: &K)
+    where
+        V: Deletable, {
+        if let Some(v) = self.remove_and_get(key) {
+            v.delete()
+        }
+    }
+
+    fn key_with_map_prefix(&self, key: &K) -> Vec<u8> {
+        let mut key_with_prefix = self.prefix.to_vec();
+        key.serial(&mut key_with_prefix).unwrap_abort();
+        key_with_prefix
+    }
+}
+
+impl<'a, K, V, S: HasStateApi> Drop for StateMapIter<'a, K, V, S> {
+    fn drop(&mut self) {
+        // Delete the iterator to unlock the subtree.
+        if let Some(valid) = self.state_iter.take() {
+            self.state_api.delete_iterator(valid);
+        }
+    }
+}
+
+impl<K, V, S> StateMap<K, V, S>
+where
+    S: HasStateApi,
+{
+    pub(crate) fn open(state_api: S, prefix: [u8; 8]) -> Self {
+        Self {
+            _marker_key: PhantomData,
+            _marker_value: PhantomData,
+            prefix,
+            state_api,
+        }
+    }
+
+    /// Get an iterator over the key-value pairs of the map. The iterator
+    /// returns values in increasing order of keys, where keys are ordered
+    /// lexicographically via their serializations.
+    pub fn iter(&self) -> StateMapIter<'_, K, V, S> {
+        match self.state_api.iterator(&self.prefix) {
+            Ok(state_iter) => StateMapIter {
+                state_iter:       Some(state_iter),
+                state_api:        self.state_api.clone(),
+                _lifetime_marker: PhantomData,
+            },
+            Err(StateError::SubtreeWithPrefixNotFound) => StateMapIter {
+                state_iter:       None,
+                state_api:        self.state_api.clone(),
+                _lifetime_marker: PhantomData,
+            },
+            _ => crate::trap(),
+        }
```

Comment:
I am not in favor of this since it will hide any bugs or incorrect uses of the library. We want to fail hard in case we get an unexpected error, since that is either a bug in the library, or the user using low-level state.